kpi/deployment_backends/kobocat_backend.py

```diff
@@ -761,6 +770,116 @@ def set_validation_statuses(self, data, user, method):
+            )
+            deprecated_id.text = instance_id.text
+            instance_id.text = uuid_formatted
+
+            # If the form has been updated with new fields and earlier
+            # submissions have been selected as part of the bulk update,
+            # a new element has to be created before a value can be set.
+            # However, with this new power, arbitrary fields can be added
+            # to the XML tree through the API.
+            for k, v in payload.items():
+                # A potentially clunky way of taking groups and nested groups
+                # into account when the elements don't exist on the XML tree
+                # (which could be the case if the form has been updated). They
+                # are iteratively attached to the tree since we can only
+                # append one element deep per iteration
+                if '/' in k:
+                    accumulated_elements = []
+                    for i, element in enumerate(k.split('/')):
+                        if i == 0:
+                            ET.SubElement(xml_parsed, element)
+                            accumulated_elements.append(element)
+                        else:
+                            updated_xml_path = '/'.join(accumulated_elements)
+                            ET.SubElement(
+                                xml_parsed.find(updated_xml_path), element
+                            )
+                            accumulated_elements.append(element)
+
+                element_to_update_or_new = (
+                    xml_parsed.find(k)
+                    if xml_parsed.find(k) is not None
+                    else ET.SubElement(xml_parsed, k)
+                )
+                element_to_update_or_new.text = v
+
+            # TODO: Might be worth refactoring this as it is also used when
+            # duplicating a submission
+            file_tuple = (_uuid, io.BytesIO(ET.tostring(xml_parsed)))
+            files = {'xml_submission_file': file_tuple}
+            # `POST` is required by OpenRosa spec https://docs.getodk.org/openrosa-form-submission
+            kc_request = requests.Request(
+                method='POST', url=self.submission_url, files=files
+            )
+            kc_response = self.__kobocat_proxy_request(
+                kc_request, user=self.asset.owner
+            )
+            response_codes.append(kc_response.status_code)
+
+        successful_updates = response_codes.count(status.HTTP_201_CREATED)
+
+        # TODO: This follows a similar reponse message that bulk updating of
+        # `validataion_statuses` will give, however there may be something more
+        # useful to send back to the client
+        if successful_updates > 0:
+            return {
+                'status': status.HTTP_200_OK,
+                'data': {
+                    'detail': _(
+                        '{} submissions of {} attempts have been updated'
```

Comment:
It's passed to a `Response` object in the data viewset and expanded following the same pattern I saw implemented there: `return Response(**json_response)` — where it must be JSONified. The other methods in `KobocatDeploymentBackend` typically do something like `return self.__prepare_as_drf_response_signature(...)` and `__prepare_as_drf_response_signature` returns a dict to the viewset hence following that pattern. I'm happy to change that though if we're wanting to move away from that.

Reply:
sorry for being unclear: i was specifically referring to the string `{} submissions of {} attempts have been updated`. if we expect a machine to glean something useful from that, we're not being too nice to our fellow developers by asking them to do string parsing to get the two integers out in order to see whether the operation completed (totally) successfully. 

i did miss earlier the comment that says this was inspired by the way bulk updating validation statuses works (hey, btw, there is a `validataion` typo in the comment :wink:). is [this](https://github.com/kobotoolbox/kobocat/blob/d91834ccb174da0fd0305f8e8ae11ed93cce10c0/onadata/apps/api/viewsets/data_viewset.py#L443-L446) the response you saw? if so, i can see that i'm not evenly applying my string-parsing criticism :)

i would draw one point of distinction here, though: it doesn't appear that bulk updates of validation statuses can fail partially. the `{} submissions have been updated` message is an (admittedly not machine-friendly) informational accompaniment to a HTTP 200 response that always means total success.

i did glance at the ~~HTTP spec~~, er, [someone's interpretation of the HTTP spec](https://stackoverflow.com/questions/8472935/http-status-code-for-a-partial-successful-request#comment114923447_22911940) when i did the previous review and concluded that 200 response was just fine for partial success. trying to think, though, about what the front end would need to do when it receives a response made me wonder if something like `{successes: 7, failures: 3}` might be more helpful than `7 submissions of 10 attempts have been updated`. i did a little searching and can't find a good example of best practices for this. since we're setting a precedent (for ourselves) here, it might be worth a quick chat in Zulip.

Reply:
Thanks for the detailed explanation (and catching that sneaky typo) — that makes a lot of sense 💪 The changes have been made according to our discussion on Zulip.